framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,305 @@
+#include <unistd.h>
+#include <assert.h>
+#include <media/HttpInputDataSource.h>
+#include <chrono>
+
+#include "utils/MediaUtils.h"
+#include "HttpStream.h"
+#include "StreamBuffer.h"
+#include "StreamBufferReader.h"
+#include "StreamBufferWriter.h"
+
+
+#ifndef CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_SIZE
+#define CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_SIZE 4096
+#endif
+
+#ifndef CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD
+#define CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD 2048
+#endif
+
+#ifndef CONFIG_HTTPSOURCE_DOWNLOAD_STACKSIZE
+#define CONFIG_HTTPSOURCE_DOWNLOAD_STACKSIZE 8192
+#endif
+
+namespace media {
+namespace stream {
+
+// Content-Type tag
+static const std::string TAG_CONTENT_TYPE = "Content-Type:";
+
+static const std::chrono::seconds WAIT_HEADER_TIMEOUT = std::chrono::seconds(3);
+static const std::chrono::seconds WAIT_DATA_TIMEOUT = std::chrono::seconds(3);
+
+HttpInputDataSource::HttpInputDataSource(const std::string &url)
+	: InputDataSource(), mUrl(url), mThread((pthread_t)0)
+{
+	medvdbg("url: %s\n", mUrl.c_str());
+}
+
+HttpInputDataSource::HttpInputDataSource(const HttpInputDataSource &source)
+	: InputDataSource(source), mUrl(source.mUrl), mThread((pthread_t)0)
+{
+}
+
+HttpInputDataSource &HttpInputDataSource::operator=(const HttpInputDataSource &source)
+{
+	InputDataSource::operator=(source);
+	return *this;
+}
+
+bool HttpInputDataSource::open()
+{
+	medvdbg("HttpInputDataSource::open!\n");
+
+	if (isPrepare()) {
+		medvdbg("HttpInputDataSource is already opened!\n");
+		return true;
+	}
+
+	if (mStreamBuffer == nullptr) {
+		mStreamBuffer = StreamBuffer::Builder()
+								.setBufferSize(CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_SIZE)
+								.setThreshold(CONFIG_HTTPSOURCE_DOWNLOAD_BUFFER_THRESHOLD)
+								.build();
+
+		if (mStreamBuffer == nullptr) {
+			meddbg("mStreamBuffer is nullptr!\n");
+			return false;
+		}
+
+		mStreamBuffer->setObserver(this);
+		mBufferReader = std::make_shared<StreamBufferReader>(mStreamBuffer);
+		mBufferWriter = std::make_shared<StreamBufferWriter>(mStreamBuffer);
+	}
+
+	if (mHttpStream == nullptr) {
+		mHttpStream = HttpStream::create();
+		if (mHttpStream == nullptr) {
+			meddbg("mHttpStream is nullptr!\n");
+			return false;
+		}
+	}
+
+	std::unique_lock<std::mutex> lock(mMutex);
+	mIsHeaderReceived = false;
+	mIsDataReceived = false;
+
+	pthread_attr_t attr;
+	pthread_attr_init(&attr);
+	pthread_attr_setstacksize(&attr, CONFIG_HTTPSOURCE_DOWNLOAD_STACKSIZE);
+	struct sched_param sparam;
+	sparam.sched_priority = 100;
+	pthread_attr_setschedparam(&attr, &sparam);
+
+	int iRet = pthread_create(&mThread, &attr, static_cast<pthread_startroutine_t>(workerMain), this);
+	if (iRet != OK) {
+		meddbg("Fail to create download thread, err:%d\n", iRet);
+		return false;
+	}
+	pthread_setname_np(mThread, "HttpSourceDownloader");
+
+	// wait for Content-Type header
+	if (!mCondv.wait_for(lock, WAIT_HEADER_TIMEOUT, [=]{ return mIsHeaderReceived; })) {
+		medwdbg("download:: wait header timeout!\n");
```

Comment:
Warning is correct?

Reply:
It's an error log, I corrected it. `medwdbg` -> `meddbg`